### ISSUE TITLE
torch distributed training with multi gpus errors in GRPOtrainer

### ISSUE THREAD

**Issue by jinhonglu** [huggingface/trl#3451]
### Reproduction

```
python -m torch.distributed.launch --nproc_per_node=1 --nnodes=1 --node_rank=0 --master_addr
=127.0.0.1 --use_env --master_port=12345 train.py train.training_arguments.output_dir=outputs/test_grpo
```

I am fine-tuning Qwen2-0.5b model with GRPO, I successfully launched the training with 1 GPU. However, the training was failed when I tried to launch with multiple gpus.

The error shows that the linear layer mat2 requires 2-d matrix, not 1-d matrix.

I had checked the inputs to the model in both 1 GPU and multi-GPUs training are the same. I wonder what causes this error?

### System Info

python3.9
torch                    2.6.0
transformers             4.51.3
trl                      0.17.0

h100 80G

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by xiaohaochen0308** [comment]
我也是，可否交流一下

**Comment by xiaohaochen0308** [comment]
19856014791
微信交流可否



发自我的iPhone


------------------ Original ------------------
From: Jiko ***@***.***&gt;
Date: Thu,May 15,2025 9:25 PM
To: huggingface/trl ***@***.***&gt;
Cc: Xiaohao Chen ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [huggingface/trl] torch distributed training with multi gpuserrors in GRPOtrainer (Issue #3451)

**Comment by shirinyamani** [comment]
@jinhonglu @xiaohaochen0308 
Please use English, as it can benefit larger community!

**Comment by zaiquanyang** [comment]
> 19856014791
> 微信交流可否
> 
> 
> 
> 发自我的iPhone
> […](#)

Do you have solved that? I met the similar problem as https://github.com/huggingface/trl/issues/3035

**Comment by Mike-7777777** [comment]
What is your reward function?

**Comment by jinhonglu** [comment]
> What is your reward function?

Is this related to reward function?

**Comment by albertvillanova** [comment]
Hey @jinhonglu, sorry for the very late reply — we're catching up on the backlog. 

The report doesn't include a runnable reproduction: the `train.py` your command launches isn't shared, and the only code shown (`Qwen2ForCausalLM.from_pretrained(...).state_dict()`) is your own model-loading rather than a GRPOTrainer setup, so there's nothing we can run to reproduce. 

That said, `mat2 must be a matrix, got 1-D tensor` showing up only in multi-GPU runs is the classic DeepSpeed ZeRO-3 symptom — under ZeRO-3 each rank holds parameters partitioned/flattened to 1-D, so pulling a raw `.state_dict()` (or matmul-ing the weights) outside a gathered context yields 1-D tensors; TRL handles this internally via `ds3_gather_for_generation` / `deepspeed.zero.GatheredParameters` (see https://github.com/huggingface/trl/blob/0ed4ac7fa03bf388a94566f1b1f71e71445ca6e7/trl/trainer/grpo_config.py#L426-L433), but custom loading code in your own script won't. Also drop `device_map="auto"` for distributed/ZeRO runs. 

I'm closing this for now, but please re-open with a minimal self-contained script (your `GRPOConfig` + how you build the model) and the full traceback, and we'll dig in.
